flink/src/main/scala/ru/itclover/tsp/services/ArrowService.scala

```diff
@@ -0,0 +1,137 @@
+    * @return tuple with schema and reader
+    */
+  def retrieveSchemaAndReader(inputData: Array[Byte]): (Schema, ArrowReader, BaseAllocator) = {
+
+    val allocator = new RootAllocator(Integer.MAX_VALUE)
+    val bytesStream = new ByteArrayInputStream(inputData)
+
+    val reader = new ArrowStreamReader(bytesStream, allocator)
+
+    (reader.getVectorSchemaRoot.getSchema, reader, allocator)
+
+  }
+
+  /**
+  * Method for retrieving schema and reader from input file
+    * @param inputData file with input data
+    * @return tuple with schema and reader
+    */
+  def retrieveSchemaAndReader(inputData: File): (Schema, ArrowReader, BaseAllocator) = {
+
+    val allocator = new RootAllocator(Integer.MAX_VALUE)
+    val fileStream = new FileInputStream(inputData)
+    val readChannel = new SeekableReadChannel(fileStream.getChannel)
+
+    val reader = new ArrowFileReader(readChannel, allocator)
+
+    (reader.getVectorSchemaRoot.getSchema, reader, allocator)
+
+  }
+
+  /**
+  * Method for converting input to list of values
+    * @param input arrow schema, arrow reader, arrow allocator
+    * @return list of values
+    */
+  def convertData(input: (Schema, ArrowReader, BaseAllocator)): ListBuffer[ListBuffer[(String, Any)]] = {
+
+    val schema = input._1
+    val reader = input._2
+    val allocator = input._3
+
+    val schemaFields: List[String] = schema.getFields
+                                           .asScala
+                                           .map(_.getName)
+                                           .toList
+
+    val schemaRoot = reader.getVectorSchemaRoot
+    var rowCount = 0
+
+    var readCondition = reader.loadNextBatch()
+
+    var result: mutable.ListBuffer[mutable.ListBuffer[(String, Any)]] = mutable.ListBuffer.empty
+
+    while(readCondition){
+
+      rowCount = schemaRoot.getRowCount
+
+      for(i <- 0 until rowCount){
+
+        var rowResult: mutable.ListBuffer[(String, Any)] = mutable.ListBuffer.empty
+
+        for(field <- schemaFields){
+
+          val valueVector = schemaRoot.getVector(field)
+
+          if(!typesMap.contains(valueVector.getMinorType)){
+            throw new IllegalArgumentException(s"There is no mapping for Arrow Type ${valueVector.getMinorType}")
+          }
+
+          val valueInfo = typesMap(valueVector.getMinorType)
+
+          val transferredVector: BaseValueVector = valueInfo.getName match {
+            case "int" => valueVector.asInstanceOf[IntVector]
+            case "boolean" => valueVector.asInstanceOf[BitVector]
+            case "java.lang.String" => valueVector.asInstanceOf[VarCharVector]
+            case "float" => valueVector.asInstanceOf[Float4Vector]
+            case "double" => valueVector.asInstanceOf[Float8Vector]
+            case "long" => valueVector.asInstanceOf[BigIntVector]
+            case "short" => valueVector.asInstanceOf[SmallIntVector]
+            case _ => throw new IllegalArgumentException(s"No mapper for type ${valueInfo.getName}")
```

Comment:
throwing an exceptions is a side-effect, better to avoid it returning Either

Reply:
done